packages/xforms-engine/src/lib/reactivity/validation/createAggregatedViolations.ts

```diff
@@ -0,0 +1,61 @@
+import { createMemo } from 'solid-js';
+import type {
+	AncestorNodeValidationState,
+	DescendantNodeViolationReference,
+} from '../../../client/validation.ts';
+import type { AnyParentNode, AnyValueNode } from '../../../instance/hierarchy.ts';
+
+const violationReference = (node: AnyValueNode): DescendantNodeViolationReference | null => {
+	const violation = node.getViolation();
+
+	if (violation == null) {
+		return null;
+	}
+
+	const { nodeId } = node;
+
+	return {
+		nodeId,
+		get reference() {
+			return node.currentState.reference;
+		},
+		node,
+		violation,
+	};
```

Comment:
This was kind of a kitchen sink design, because I didn't have a clear picture of how the client implementation will use these aggregate references to address UX we're targeting.

My thinking actually goes the opposite direction: how much do clients need/want direct access to the node here? If one of the other references is suitable, I'd prefer to _eventually remove_ `node` from the API. That would greatly simplify approaches to maintaining this shared reactive state (and potentially evolving it to be lazy).